lib/Dialect/AIEX/Transforms/AIEDmaToNpu.cpp

```diff
@@ -419,6 +419,46 @@ struct DmaWaitToNpuPattern : OpConversionPattern<NpuDmaWaitOp> {
   }
 };
 
+std::optional<AIE::ShimDMAAllocationOp>
+getAllocOpForSymbol(SmallVector<AIE::ShimDMAAllocationOp> shimDmaAllocOps,
+                    StringRef sym_name) {
+  for (auto shimDmaAllocOp : shimDmaAllocOps)
+    if (shimDmaAllocOp.getSymName() == sym_name)
+      return shimDmaAllocOp;
+  return std::nullopt;
+}
+
+void insertNpuSyncOpForResults(AIE::DeviceOp device) {
+  SmallVector<AIE::ShimDMAAllocationOp> shimDmaAllocOps;
+  device.walk([&](AIE::ShimDMAAllocationOp shimDmaAllocOp) {
+    shimDmaAllocOps.push_back(shimDmaAllocOp);
+  });
+  device.walk([&](mlir::func::FuncOp f) {
+    SmallVector<AIEX::NpuDmaMemcpyNdOp> dmas;
+    Operation *returnOp = nullptr;
+    f.walk([&](mlir::func::ReturnOp op) { returnOp = op.getOperation(); });
+    f.walk([&](AIEX::NpuDmaMemcpyNdOp dma) { dmas.push_back(dma); });
+    for (auto dma : dmas) {
+      if (auto infoOp =
+              getAllocOpForSymbol(shimDmaAllocOps, dma.getMetadata())) {
+        if (infoOp->getChannelDir() == AIE::DMAChannelDir::S2MM) {
+          // Found dma op copying results to host
+          OpBuilder builder(dma);
+          auto col = builder.getI32IntegerAttr(infoOp->getCol());
+          auto row = builder.getI32IntegerAttr(0);
+          auto dir = builder.getI32IntegerAttr(0);
+          auto chan = builder.getI32IntegerAttr(infoOp->getChannelIndex());
+          auto col_num = builder.getI32IntegerAttr(1);
+          auto row_num = builder.getI32IntegerAttr(1);
+          builder.setInsertionPoint(returnOp);
+          builder.create<AIEX::NpuSyncOp>(dma->getLoc(), col, row, dir, chan,
+                                          col_num, row_num);
+        }
+      }
+    }
```

Comment:
I don't think we want to unconditionally add a sync to every outgoing memcpy. For example, what if we are collecting N output tiles at the shim and only need to sync at the end? The N-1 extraneous syncs will have a performance penalty vs. the single (manually inserted) sync at the end.